misc/gam/gam_one_centered_design.stan

```diff
@@ -0,0 +1,177 @@
+    N <- rows(X);
+    K <- cols(X);
+    {
+      //form QR decomposition of Xt1
+      matrix[K,1] sums;
+      vector[K] u;
+      sums <- to_matrix(rep_row_vector(1,N)*X)';
+      return householder(sums)[,1];
+    }
+  }
+  
+  //apply centering constraint on X to a matrix D: D -> DZ
+  // where Z is such that X^T 1 = QR with Q=[u:Z]
+  // if D is LxK, and X is NxK, returns a Lx(K-1) matrix
+  matrix center(matrix X, matrix D) {
+    vector[cols(X)] u;
+    u <- centering_constraint(X);
+    return D[,2:] - (2*D*u)*u[2:]';
+  }
+  
+  //difference penalty for P-splines
+  //K is the size of the parameter vector
+  //d is the order of the difference
+  //Returns P, which is K-d x K
+  //D=t(P)P is the difference matrix, P*beta is the difference vector
+  matrix difference_matrix_sqrt(int K, int d) {
+    matrix[K,K] P;
+    P <- diag_matrix(rep_vector(1, K));
+    for (i in 1:d) {
+      matrix[K-i,K] tmp; //stan 2.9.0 issues compile error for deep_copy
+      tmp <- P[2:(K-i+1),]-P[1:(K-i),];
+      P[1:(K-i),] <- tmp;
+    }
+    return P[1:(K-d),];
+  }
+
+  // design matrix for 1D b-spline fit of degree q with K functions
+  matrix design(vector x, int K, int q) {
+    matrix[rows(x),K] X;
+    X <- bspline(x, K, q);
+    //intercept and bspline basis: K params
+    return append_col(rep_vector(1,rows(x)), center(X,X));
+  }
+  
+  // difference operator for that design matrix
+  matrix difference_op(vector x, int K, int q, int d) {
+    matrix[K-d,K-1] diff;
+    diff <- center(bspline(x, K, q), difference_matrix_sqrt(K,d));
+    return append_col(rep_vector(0,K-d), diff); //intercept has no constraint
+  }
+}
+////////////////////////////////////////////////////////////////
+data {
+  int<lower=1> N; //number of data points
+  int<lower=1> K; //number of parameters
+  int y[N]; //dependent variable
+  vector[N] x; //independent variable
+}
+transformed data {
+  matrix[N,K] X; //design matrix
+  matrix[K-difforder(),K] P; //difference operator
+  X <- design(x, K, splinedegree());
+  P <- difference_op(x, K, splinedegree(), difforder());
+}
+parameters {
+  vector[K] beta; //parameter vector for intercept + centered spline
+  real<lower=0> alpha; //precision of negative binomial
+  real<lower=0> lambda; //regularization parameter
+}
+model {
+  //negative binomial GAM
+  y ~ neg_binomial_2(exp(X * beta), alpha);
+  //P-spline prior on the 2nd order differences (K-2 params)
+  //warning on jacobian can be ignored
+  //see GAM, Wood (2006), section 4.8.2 (p.187)
+  P*beta ~ normal(0, 1./(alpha*lambda));
+}
+generated quantities {
+  matrix[N,K] designmat; //design matrix
+  matrix[N,K] weighted; //weighted basis functions
+  vector[N] pred; //spline interpolant
+  real edf; //effective degrees of freedom
+  vector[K] edfvec;
+  designmat <- X;
+  weighted <- X .* rep_matrix(beta', rows(x));
+  pred <- exp(X * beta);
+  {
+    matrix[K,K] XtX;
+    XtX <- crossprod(X);
+    edfvec <- diagonal(inverse_spd(XtX+lambda*crossprod(P)) * XtX);
```

Comment:
You don't ever want to do inverses if you can avoid them.  So rather than inv(a) * b, you want `a \ b`.  But then I don't think we have an spd-specific version (symmetric pos definite).

And we'd also prefer spaces between `N,K` as in `N, K`.